### DISCUSSION TITLE
Serial PCM Audio to and from AnalogAudioStream ESP32

### DISCUSSION THREAD

**Discussion by wsciaroni** [pschatzmann/arduino-audio-tools#1769] [Q&A] [answered]
I am have a project that I am trying to run on the ESP32. I have a couple of fundamental pain points as I learn to use this library.

### Project

You can take a look at my embedded code for this project here: https://github.com/wsciaroni/wsciaroni_kv4p-ht/blob/d6fd7c3e870ea084421675657cef2f4eceb9f21a/wsciaroni_kv4p-ht.ino

Currently, I have a GeneratedStream just to put audio through, but that is just for testing.

### Project Requirements

- **Hardware**, I am using the esp32 devkit v1. I am using its built-in ADC and DAC.
- **Audio Sample Rate**: The audio should have a sample rate of 44100Hz.
- **Audio Channels**: I only need one audio channel. The audio is mono.
- **Serial Settings**: I want Serial to be configured with a baud rate of 921600.
- **Serial Protocol**: Simple command/data protocol with length indication. (Described below)

I want two modes for this. Either rx or tx.

In RX mode, I want to use the AudioTools library to read from the built in Analog to Digital converter. I want to use the AudioTools library to automatically send this audio stream to the serial port in 8 Bit PCM format. No separators.

In TX mode, I want to manually add bytes to an AudioTools BufferedAudioStream. I want to use the AudioTools library to automatically send this audio stream to the built in Digital to Analog converter. I will put the bytes on the stream as I read them from 8 BitPCM coming in from Serial.

When in Tx mode, the rx audio should not be sent.

The Serial communication will come in the following format:

- Other device will send one byte indicating whether the following is a CMD or DATA.
- If it is a CMD, one more byte will be provided indicating which mode to change to.
- If it is a DATA, two bytes will be provided indicating the length of the data to come. Then, the data will be the tx audio stream in 8 bit PCM format.

I think I want to use an `AnalogAudioStream` to read from and write to the ADC and DAC.
I think I want to use StreamCopy to take data from the incoming stream and write to the outgoing stream.

I don't think I can begin the AnalogAudioStream In in RX_MODE and AnalogAudioStream Out int TX_MODE at the same time due to the underlying implementation.

### Questions

1. How should I most efficiently read a chunk of PCM8bit data from Serial and write it to a buffer then have the data written via the internal DAC?

2. How should I most efficiently read from the internal ADC and write the data to Serial in a PCM8bit stream of bytes.

3. What is the setup and teardown required to start and stop the different streams when I switch modes?

**Comment by pschatzmann** [comment, marked as answer]
You would use an EncodedAudioStream with an 8 bit codec!
You can directly use Serial as source and destination...

I think your sample rate is a bit high for Serial:  you need to use > 396900 baud

I think there are some examples in the communction folder!

**Reply by wsciaroni** [reply]
I'm not sure how I could use an EncodedAudioStream for the TX Mode since the Serial handles both my control signals and the pcm8 audio. What would you suggest?

<Edit: I see the example on 8 bit for my Rx mode>

Also, are you suggesting I turn up the Serial baud rate to `396900`? I am using Serial baud rate of `921600` which should be greater than the `44100*8 = 396900` required, right?

**Reply by wsciaroni** [reply]
Thank you for the quick reply!

**Reply by wsciaroni** [reply]
Also this one: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-communication/serial/send-8bit-receive/send-8bit-receive.ino

I think this example solves my question <2> from my original post. However, I'm still stuck on the others:

How should I most efficiently read a chunk of PCM8bit data from Serial and write it to a buffer then have the data written via the internal DAC?

What is the setup and teardown required to start and stop the different streams when I switch modes? I don't think I can set the ADC to be both RX and TX simultaneously can I?

**Reply by pschatzmann** [reply]
For your first open question: you don't: let the framework handle this for you! Read the chapter about Encoding/Deconding. You want to use L8!

Just follow the Arduino conventions and call end() ! If you have any doubts check in the Class documentation for the available methods.

**Reply by wsciaroni** [reply]
Maybe I left some details out of my write-up.

**The Serial in to the ESP32 contains a simple command/data protocol with length indication**. This means that I cannot use the EncodedAudioStream to read from serial (or **I would consume my CMD messages as audio data**!).

How would you propose using your framework for me to indicate I have a certain number of bytes of audio data available on Serial without the stream consuming my CMD messages?

**Reply by pschatzmann** [reply]
Use a container: I have the BinaryContainer which does what you want: your commans you can send as metadata.
But I guess it would be more reliable if you just use a separate serial port for the commands.

**Reply by wsciaroni** [reply]
Okay, I read over https://www.pschatzmann.ch/home/2023/04/28/arduino-audio-tools-codecs-and-containers/

It looks like it requires a fixed length MetaData? Is that correct? It shouldn't be that big of a deal to make that happen. So, this keys off of the `ContainerType` when it reads its header? Then it determines if the incoming is Metadata or Data?

I don't see an example of `pure audio pcm data along with some additional custom (meta) information`. Is that correct?

I would love to have been able to use a separate Serial connection for ctrl vs audio. Unfortunately, I can't use another Serial connection (The other one is already in use to control a different chip).

I'm open to suggestions though. This is for an App to send and receive data on androids over USB.

**Comment by pschatzmann** [comment]
You can use any pin with SoftwareSerial
If I remember right, the meta data is variable length as well: therefore you you need to specifiy the length when writing data
and you can add it anytimes into the stream.

**Reply by wsciaroni** [reply]
Great stuff! I'll look into this and come back when I know enough to be dangerous.

**Reply by wsciaroni** [reply]
I have attempted to implement the method you stated: https://github.com/wsciaroni/wsciaroni_kv4p-ht/blob/SwitchToPschatzmannMethod/wsciaroni_kv4p-ht.ino

Could you please provide some feedback on this?

What I think needs to be sent from the other device in order to invoke the Metadata callback:

For a simple command:
`{ (byte)'\r', (byte)'\n', Meta.getByte(), 0, 0, 0, command.getByte() }`

For a command with params:
`{ (byte)'\r', (byte)'\n', Meta.getByte(), 0, 0, 0, command.getByte(), ...(bytes with MetaData) }`

For audio:
`{ (byte)'\r', (byte)'\n', Audio.getByte(), (byte)(numberOfBytesToSend >> 8), (byte)numberOfBytesToSend, (byte)0, ...(each byte 
 for i = 0; i < numberOfBytesToSend) }`

https://github.com/wsciaroni/wsciaroni_kv4p-ht/blob/fd47dda585f68b9b68c7cbb56a7e65dce18759e5/wsciaroni_kv4p-ht.ino#L154

**Reply by pschatzmann** [reply]
I don't quite understand: you would just call writeMetaData() with whatever you want to send.

https://www.pschatzmann.ch/home/2023/04/28/arduino-audio-tools-codecs-and-containers/

On the receiving side you just register a callback that should receive your meta data.

**Reply by wsciaroni** [reply]
I'm sending the data from a java app, so I can't call writeMetadata ... I'm looking to understand the way that class expects the serial to be written to trigger the callback for the decoder.

**Reply by pschatzmann** [reply]
Oh, I see: so you need to convert BinaryContainerEncoder to java or roll your own solution. 
In the BinaryContainerEncoder each record  starts with a CRLF for readability followed by  the container type, the  length (of the header + data) and a checksum.

To come back to your question about reading and writing, I suggest you read the first chapter in the Wiki!

**Reply by wsciaroni** [reply]
Would you still suggest using the container?

I'm thinking it would be more straightforward to manually write the data from serial into a buffered encoded stream. I'm curious to know your thoughts on this.

I don't see any examples of manually writing to a stream, but I know this is an edge case.

**Reply by pschatzmann** [reply]
I don't see any problem with this: just read the [The most important basics!](https://github.com/pschatzmann/arduino-audio-tools/wiki/Introduction) Wiki

**Reply by wsciaroni** [reply]
Yes, I have made heavy use of this documentation.

> I don't see any problem with this

You don't see a problem with me writing the bytes to the stream manually? If so, what would that look like in terms of Stream configuration? I cannot get it to line up.

Please advise.

**Reply by pschatzmann** [reply]
I don't quite understand your question: What do you mean with line up ?

**Reply by wsciaroni** [reply]
I cannot find an example of how to configure an audio stream such that I can write 8 bit PCM to the stream manually with the `write` function after reading it from Serial.

**Reply by pschatzmann** [reply]
No special configuration is necessary: by calling write() and readBytes() you just do the same that I actually do in the StreamCopy.

**Reply by wsciaroni** [reply]
I'm running into something that I don't know how to troubleshoot, so I am yet again reaching out for your thoughts about [my sketch](https://github.com/wsciaroni/wsciaroni_kv4p-ht/blob/cdbdaa99f46374265c95b5ee5d373ee11ae06a19/wsciaroni_kv4p-ht.ino)

## This works

```c++
AudioInfo info(44100, 1, 16);

// AnalogAudioStream in;

SineWaveGenerator<int16_t> sineWave(32000);
GeneratedSoundStream<int16_t> in(sineWave);

auto &serial = Serial;
EncoderL8 enc;
EncodedAudioStream enc_stream(&serial, &enc);
StreamCopy copierOut(enc_stream, in);

void setup()
{
    auto config = in.defaultConfig(RX_MODE);
    config.copyFrom(info);
    config.use_apll = true;

    sineWave.begin(info, N_B4);
    in.begin(config);

    enc_stream.begin(info);

    copierOut.begin();
}

void loop()
{
  copier.copy();
}
```

## This does not work

I hear really distorted audio that sounds awfull.

```c++
AudioInfo info(44100, 1, 16);

AnalogAudioStream in;

auto &serial = Serial;
EncoderL8 enc;
EncodedAudioStream enc_stream(&serial, &enc);
StreamCopy copierOut(enc_stream, in);

void setup()
{
    auto config = in.defaultConfig(RX_MODE);
    config.copyFrom(info);
    config.use_apll = true;

    in.begin(config);

    enc_stream.begin(info);

    copierOut.begin();
}

void loop()
{
  copier.copy();
}
```

## Other Info

I can provide the i2s settings I was using before attempting to migrate to your library if needed.

**Reply by pschatzmann** [reply]
I guess you need some flow control to make sure that you don't loose any data

**Reply by wsciaroni** [reply]
Why would it work for the Sin and not for the Analog when they are both configured with the same `AudioInfo`?

**Reply by pschatzmann** [reply]
Your sketch from which you pretend it is working, can't work: it will give compile errors.
The second one as well...

The different between GeneratedSoundStream and AnalogAudioStream is the speed with which the audio is generated: the first does it as fast as possible, the second by the defined rate!

The second is tricky when you have multiple systems involved because the clocks are not synchronized: usually filling a big buffer on the receiver side before starting to play will help

**Reply by wsciaroni** [reply]
That's a verry sharp eye. If you had looked at the link to the full code, you would have seen the full sketch. I simplified the code for you to see the change. If you would look at the full sketch, you would see my `AUDIO_USE_SIN_FOR_TESTING` that I use to switch between the sin and analog. With only that change, I get broken audio.

**Reply by pschatzmann** [reply]
see the second part of my answer above

**Reply by wsciaroni** [reply]
Thanks for your help! I appreciate you taking the time to look at this.

I've already implemented a fairly large buffer on the receiving end, longer than I've used before I switched to AudioTools. I'm starting to suspect it might be related to the ADC resolution. I noticed the audio improves when the signal to the ADC is louder.

I was using `ADC-ATTEN_DB_0` before, but I'm not sure how this is being set (I couldn't find it in the docs). I saw how to change it with the ESP32 v3.0 or newer, but I'm using ESP32 v2.0.17. Would you recommend changing to the newer board or sticking with v2? I saw there were large changes in the ADC/DAC driver (via i2s ADC removal from the IDF??). Do you have any suggestions for configuring the ADC attentuation in v2.0.17?

**Reply by pschatzmann** [reply]
My recommendation would be to use the latest ESP32 Arduino core version.
[Analyse the audio from the ADC with the Serial Plotter](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working)!
